### PR TITLE
Prac: Intro to RL

### PR DESCRIPTION
**Pull request by jcformanek**
I have implemented an introductory practical on RL. There are three main sections. One on Random Policy Search, a second on Policy Gradients (REINFORCE), and a final one on Q-Learning (DQN). 

The explanations and exercises in the first two sections are mostly complete. But the Q-Learning section is still lacking.

Comments and contributions are very welcome.

### CODE REVIEW

**Comment by KaleabTessera** [comment]
Colab link - https://colab.research.google.com/github/deep-learning-indaba/indaba-pracs-2022/blob/prac_intro_to_rl/intro_to_rl.ipynb.

File: intro_to_rl.ipynb [hunk 1]
```diff
@@ -0,0 +1,3225 @@
+{
```
Comment (sbodenstein) [inline intro_to_rl.ipynb]: <ol><li>stocastic -&gt; stochastic</li><li>Comma after <code>on some observation o_t</code></li></ol>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='16'/>

Comment (sbodenstein) [inline intro_to_rl.ipynb]: <ol><li>Lots of inconsistencies with JAX versus jax. Also JAX Numpy should be JAX NumPy, but prefer <code>jax.numpy.where</code> or <code>jax.lax.select</code>.</li><li>I also think introducing <code>jax.lax.select</code> here is cryptic, as the doc page you are linking to only tells you that it "Wraps XLA’s Select operator." </li><li>stocastic -&gt; stochastic.</li><li>Usefull -&gt; Useful</li></ol>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='19'/>

Comment (sbodenstein) [inline intro_to_rl.ipynb]: <p>Nit: there is a lot of inconsistent spacing with function args. Either a space or no space after comma.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (sbodenstein) [inline intro_to_rl.ipynb]: <p>I don't think its made clear why the if statement answer is wrong or bad. Pedagogically, it is far clearer than the jittable JAX example. At the very least, if you are going for jittable jax examples here, it should be made clear that this is expected. Personally, I would probably prefer just using numpy and standard python for this stuff, rather than forcing consideration of a second tricky concept (jittable jax).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='26'/>

Comment (sbodenstein) [inline intro_to_rl.ipynb]: <ul><li>I find this language rather too technical for someone seeing this for the first time. Rather than abstract discussions around terminal states, maybe bring this back to the only RL example lots of these tutorial takers know about: the cartpole. We want to maximize the time it stays up, and it gets one reward per timestep. So...</li><li>Also, I would just be confused by a lot of this: why would we want discount factors with infinite horizons etc etc (seems completely arbitrary). Is this really necessary for the first few RL methods? (random policy search doesn't care about discount factors)</li><li>expected discounted return: isn't an expectation something over a distribution? Where have you introduced a distribution, and wouldn't this be E[R_t]?</li></ul><p> </p><p>I think this section is very confusing and would remove the whole thing, or move it to after actually solving the problem (abstractions make more sense then to learners).</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='30'/>

Comment (sbodenstein) [inline intro_to_rl.ipynb]: <p>People usually use discount factors to allow proofs for MDPs by preventing infinite returns, which seems like a rather technical point at this stage (and yes you could use it for other purposes. But is this important for a first example of RL? Its much conceptually clearer just saying 'we want to keep the cartpole from falling over for as long as possible!')</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (sbodenstein) [inline intro_to_rl.ipynb]: <blockquote> the game is over and the agent receives a reward of 0</blockquote><p>Slightly confusing: could mean that the total reward it gets is zero. I would just say the game is over, and it can no longer get anymore reward.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (sbodenstein) [inline intro_to_rl.ipynb]: <p><strong>transition</strong> to <code>Transition</code> . Nit: done -&gt; <code>done</code>  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='36'/>

Comment (sbodenstein) [inline intro_to_rl.ipynb]: <p>Nit: full name is used for all variables except <code>obs</code> . I think it will be more understandable to make this into observation.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='37'/>

Comment (sbodenstein) [inline intro_to_rl.ipynb]: <p>Nit: Jax -&gt; JAX</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='44'/>

Comment (sbodenstein) [inline intro_to_rl.ipynb]: <p>This is a Haiku feature. Why not just use plain JAX random numbers? (so create a seed jax.random.PRNGKey(0) and then use <code>jax.random.split</code> inside <code>run_environment_loop</code>). Using any haiku features outside of a <code>hk.transform</code> is dangerous. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

Comment (sbodenstein) [inline intro_to_rl.ipynb]: <p>This section has become extremely complicated, for something that should take only a few lines of code. The original motivation for this section was to have a method that even fairly weak students could succeed in implementing. My feeling (given previous years students) is that this complexity is beyond most students. Some simplifications:</p><ul><li>Just use NumPy to generate new random weights. Perfectly valid and much simpler than jitting some jax weight generator.</li><li>The only jitted jax is the policy/action MLP:</li></ul><pre class="ql-syntax" spellcheck="false">@jax.jit
def policy_mlp(observation, weights):
  # logits: list of length 2
  logits = jnp.dot(observation, weights)
  return jnp.argmax(logits)
</pre><p> </p><p>All the rest is just plain Python/NumPy. The very simplest is a loop that prints the average episode return every iteration, and breaks the loop if it solves cartpole (and you have immediate access to the numpy weights). </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='50'/>

Comment (ruanvdmerwe) [reply intro_to_rl.ipynb]: <p>I agree with @sbodenstein on this. I think the timing of introducing it should be after an initial model is trained, and we now want to show more things that can be done or are done. Espicially as we are using an undiscounted return in the next exercise. </p>

File: intro_to_rl.ipynb [hunk 2]
```diff
@@ -0,0 +1,3249 @@
...
```
Comment (ruanvdmerwe) [inline intro_to_rl.ipynb]: <p>Spelling/Grammar:</p><p> </p><p>In this tutorial, we will be learning about Reinforcement Learning, a type of Machine Learning where an agent learns to <s>chose</s> choose actions in an environment that <s>lead</s> leads to maximal reward in the long run.</p><p> </p><p>Along the way you will be <s>introduces</s> introduced to some of the most important concepts and terminology in RL</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

Comment (ruanvdmerwe) [inline intro_to_rl.ipynb]: <p>Spelling grammar:</p><p> </p><p>RL algorithms will try to maximise the score in the game by <s>chosing</s> choosing the best actions</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='4'/>

Comment (ruanvdmerwe) [inline intro_to_rl.ipynb]: <p>ctrl+f "choses" and change to "chooses" :)  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='12'/>

Comment (ruanvdmerwe) [inline intro_to_rl.ipynb]: <p>Should we not print the hint here, instead of showing "hint..." perhaps? Might not be clear that hint is in a hidden docstring.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (ruanvdmerwe) [inline intro_to_rl.ipynb]: <p>Same as excersise 1, print the hint</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

File: intro_to_rl.ipynb (hunk 1 above)
Comment (KaleabTessera) [reply intro_to_rl.ipynb]: <p>We should run linters through all pracs to avoid this. </p>

Comment (KaleabTessera) [reply intro_to_rl.ipynb]: <p>I get what you guys are saying @<a href="https://github.com/sbodenstein" target="_blank">sbodenstein</a>&nbsp;and @<a href="https://github.com/ruanvdmerwe" target="_blank">ruanvdmerwe</a> .&nbsp;I think adding a concrete example, i.e. "for CartPole, episode return means we want to maximize the time it stays up..." is a great idea! </p><p> </p><p>It is important to understand Episode Return vs Reward since that is quite foundational to RL, but maybe we move the discounted return part to the beginning of Section 3 (Policy Gradients) as we get more formal. What do you think @<a href="https://github.com/jcformanek" target="_blank">jcformanek</a>&nbsp;? </p><p> </p><p>@<a href="https://github.com/sbodenstein" target="_blank">sbodenstein</a> Why wouldn't you want discount factors with infinite horizons? You typically use a discount factor for infinite horizons. (Not sure if that was a hypothetical question as potential prac attendee).</p><p> </p>

Comment (KaleabTessera) [reply intro_to_rl.ipynb]: <p>So technically, it allows the return to be bounded when using infinite horizons, but inititaive answers to this include it would allow you weight how much the agents cares about rewards in the distant future vs immediate future.&nbsp;People could think about what would happen if discount=0 vs 1. I think it is relevant question and important question. </p>

File: intro_to_rl.ipynb (hunk 2 above)
Comment (jcformanek) [reply intro_to_rl.ipynb]: <p>Resolved. Thanks.</p>

Comment (jcformanek) [reply intro_to_rl.ipynb]: <p>haha done.</p>

File: intro_to_rl.ipynb [hunk 3]
```diff
@@ -0,0 +1,3287 @@
...
```
Comment (KaleabTessera) [inline intro_to_rl.ipynb]: <p>To avoid confusion, we should probably either explain <code>evaluation</code> or remove this extra complexity. Not sure if it explained below.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-learning-indaba/indaba-pracs-2022/pull/6/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='53'/>

File: intro_to_rl.ipynb (hunk 1 above)
Comment (jcformanek) [reply intro_to_rl.ipynb]: Done.

Comment (jcformanek) [reply intro_to_rl.ipynb]: Yeah good idea. Tried to improve things where I could.

Comment (jcformanek) [reply intro_to_rl.ipynb]: <p>I have simplified this section and decided to omit the discussion about discount factors.</p>

Comment (jcformanek) [reply intro_to_rl.ipynb]: I am open to changing it. But have not done so yet.

Comment (jcformanek) [reply intro_to_rl.ipynb]: I have simplified this section now.

Comment (jcformanek) [reply intro_to_rl.ipynb]: <p>fixed this</p>

Comment (jcformanek) [reply intro_to_rl.ipynb]: <p>removed</p>

Comment (jcformanek) [reply intro_to_rl.ipynb]: <p>removed</p>

File: intro_to_rl.ipynb (hunk 2 above)
Comment (jcformanek) [reply intro_to_rl.ipynb]: <p>removed this</p>

File: intro_to_rl.ipynb (hunk 3 above)
Comment (jcformanek) [reply intro_to_rl.ipynb]: <p>removed</p>

File: intro_to_rl.ipynb (hunk 2 above)
Comment (jcformanek) [reply intro_to_rl.ipynb]: removed this section

Comment (jcformanek) [reply intro_to_rl.ipynb]: thanks

**Review by KaleabTessera — APPROVED** [review]
Epic!
